assertions/manager.go

```diff
@@ -98,61 +112,105 @@ func WithDangerousReadyToPost() Opt {
-	rollupAddr common.Address,
-	challengeManagerAddr common.Address,
+	execProvider l2stateprovider.ExecutionProvider,
 	validatorName string,
-	pollInterval,
-	assertionConfirmationAttemptInterval time.Duration,
-	stateManager l2stateprovider.ExecutionProvider,
-	postInterval time.Duration,
-	averageTimeForBlockCreation time.Duration,
-	apiDB db.Database,
+	mode types.Mode,
 	opts ...Opt,
 ) (*Manager, error) {
-	if pollInterval == 0 {
-		return nil, errors.New("assertion scanning interval must be greater than 0")
-	}
-	if assertionConfirmationAttemptInterval == 0 {
-		return nil, errors.New("assertion confirmation attempt interval must be greater than 0")
-	}
 	m := &Manager{
-		chain:                       chain,
-		apiDB:                       apiDB,
-		backend:                     backend,
-		stateProvider:               stateProvider,
-		challengeCreator:            challengeManager,
-		challengeReader:             challengeManager,
-		rollupAddr:                  rollupAddr,
-		challengeManagerAddr:        challengeManagerAddr,
-		validatorName:               validatorName,
-		pollInterval:                pollInterval,
-		confirmationAttemptInterval: assertionConfirmationAttemptInterval,
-		forksDetectedCount:          0,
-		challengesSubmittedCount:    0,
-		assertionsProcessedCount:    0,
-		postInterval:                postInterval,
-		submittedAssertions:         threadsafe.NewLruSet[common.Hash](1000, threadsafe.LruSetWithMetric[common.Hash]("submittedAssertions")),
-		averageTimeForBlockCreation: averageTimeForBlockCreation,
+		chain:                    chain,
+		apiDB:                    nil,
+		backend:                  chain.Backend(),
+		execProvider:             execProvider,
+		rollupAddr:               chain.RollupAddress(),
+		validatorName:            validatorName,
+		times:                    defaultTimings,
+		forksDetectedCount:       0,
+		assertionsProcessedCount: 0,
+		submittedAssertions:      threadsafe.NewLruSet(1000, threadsafe.LruSetWithMetric[common.Hash]("submittedAssertions")),
 		assertionChainData: &assertionChainData{
 			latestAgreedAssertion: protocol.AssertionHash{},
 			canonicalAssertions:   make(map[protocol.AssertionHash]*protocol.AssertionCreatedInfo),
 		},
 		observedCanonicalAssertions: make(chan protocol.AssertionHash, 1000),
 		isReadyToPost:               false,
 		startPostingSignal:          make(chan struct{}),
+		mode:                        mode,
+		rivalHandler:                nil, // Must be set after construction if mode > DefensiveMode
 	}
 	for _, o := range opts {
 		o(m)
 	}
+	if m.times.pollInterval == 0 {
+		return nil, errors.New("assertion polling interval must be greater than 0")
+	}
+	if m.times.confInterval == 0 {
+		return nil, errors.New("assertion confirmation attempt interval must be greater than 0")
+	}
 	return m, nil
 }
 
+// SetRivalHandler sets the rival handler for the assertion manager.
+func (m *Manager) SetRivalHandler(handler types.RivalHandler) {
```

Comment:
That's correct. In my opinion, this method and two others (`Provider.UpdateAPIDatabase`, and `Watcher.SetEdgeManager`) are small "warts" on our dependency injection story.
However, almost every dependency injection framework supports this style of "Setter Injection", so it is not like it would be a barrier to adopting such a framework in the future if we wanted to.

That being said, I had some ideas about further breaking up these cyclic dependencies:

AssertionManager <-> ChallengeManager
Idea 1: We could provide both with a channel from which the challenge manager would read and the assertion manager would write when it is posting a new rival assertion.  The downside being that we might need to make it more robust in the case of errors (maybe)
Idea 2: The challenge manager only does 4 things with the assertion manager: `Start()`, `StopAndWait()`, `LatestAgreedAssertion()`, `LayerZeroHeights(ctx)`. The first two could be moved to some higher level container responsible for managing the lifecycle of all of the long-running threads. Some dependency injection frameworks include lifecycle management features like starting and stopping threads automatically because of how common this sort of use case is.  In my opinion, `LayerZeroHeights(ctx)` probably doesn't belong in the `AssertionManager` abstraction because it's essentially just reading data that's configured on the chain, and we should be able to look it up cheaply through the chain abstraction. But, we'd probably want to add some caching logic there to keep from sending an RPC every time we want to read it.  The `AssertionManager` currently *is* that cache. But, it should be able to exist in the chain abstraction as well. This leaves only the `LatestAgreedAssertion()` method, and I *think* we could split out the reading and writing responsibilities of the `AssertionManager` such that the `ChallengeManager` would only need to depend on the `AssertionReader` or `AssertionMonitor` part of the API, which would break the cycle.

Watcher <-> ChallengeManager
Idea 1: Same as with AssertionManager, the Watcher is only telling the ChallengeManager, "Hey, I found a RoyalEdge, start a tracker for it."  This could be a channel that is provided to both objects at construction time.
Idea 2: The ChallengeManager calls 6 methods on the watcher: `Start()`, `StopAndWait()`, `HonestBlockChallengeRootEdge(ctx, assertionHash)`, `AddVerifiedHonestEdge(ctx, levelZeroEdge)`, `AllowTrackingEdgeWithParentHash(assertionHash)`, `IsSynced()`. Of course, the first two are lifecycle management and could be moved up to a higher-level abstraction. Of the other 4, 3 are read-only data lookups and only `AddVerifiedHonestEdge` modifies the internal state of the watcher.  That write could also be done over a channel (if we want.) And, the reads could be split out into some kind of `EdgeStateReader` interface that could be implemented in a separate type as well.

Provider <-> Database
My only idea here is that we might be able to make a cleaner break among the various responsibilities of the `Provider` interface so that the `Provider` doesn't actually need to implement the `HistoryCommitmentProvider` interface. This way, the database would only be needed in the BoLD side of the house and could be optionally injected during stack construction. But, actually, since it is inherently "optional" (because it is only used when the flag is passed) it's probably okay to just leave this being injected by the stack constructor using the setter.

Reply:
For now, @rauljordan and I are happy with the level of dependency injection being used in the code. Most DI frameworks support setter injection for exactly the reasons that we have introduced it into these 3 places. And, so, we're going to leave the code as is for now, secure in the knowledge that we aren't keeping ourselves from using a DI framework in the future.